artemis/db.py

```diff
@@ -60,62 +146,77 @@ def get_task_target(task: Task) -> str:
+        analysis_dict = self.task_to_dict(analysis)
+        del analysis_dict["status"]
+        if "status_reason" in analysis_dict:
+            del analysis_dict["status_reason"]
+
+        analysis = Analysis(
+            id=analysis.uid,
+            target=analysis_dict["payload"]["data"],
+            tag=analysis_dict["payload_persistent"].get("tag", None),
+            stopped=False,
+            task=analysis_dict,
+        )
+        with Session() as session:
+            session.add(analysis)
+            session.commit()
 
     def save_task_result(
         self, task: Task, *, status: TaskStatus, status_reason: Optional[str] = None, data: Optional[Any] = None
     ) -> None:
-        created_task_result = self.task_to_dict(task)
-
-        created_task_result["_id"] = created_task_result["uid"]
-        created_task_result["status"] = status
-        created_task_result["target_string"] = get_task_target(task)
-        created_task_result["status_reason"] = status_reason
-
-        # Used to allow searching in the names and values of all existing headers
-        created_task_result["headers_string"] = " ".join([key + " " + value for key, value in task.headers.items()])
-
+        to_save = dict(
+            task=self.task_to_dict(task),
+            id=task.uid,
+            analysis_id=task.root_uid,
+            status=status,
+            tag=task.payload_persistent.get("tag", None),
+            receiver=task.headers.get("receiver", None),
+            target_string=get_task_target(task),
+            status_reason=status_reason,
+            # Used to allow searching in the names and values of all existing headers
+            headers_string=" ".join([key + " " + value for key, value in task.headers.items()]),
+        )
         if isinstance(data, BaseModel):
-            created_task_result["result"] = data.dict()
+            to_save["result"] = data.dict()
         elif isinstance(data, Exception):
-            created_task_result["result"] = str(data)
+            to_save["result"] = str(data)
         else:
-            created_task_result["result"] = data
-
-        with self.client.start_session() as session:
-            with session.start_transaction():
-                result = self.task_results.update_one(
-                    upsert=True, filter={"_id": created_task_result["uid"]}, update={"$set": created_task_result}
-                )
-                if result.upserted_id:  # If the record has been created, set creation date
-                    result = self.task_results.update_one(
-                        {"_id": created_task_result["uid"]}, {"$set": {"created_at": datetime.datetime.now()}}
-                    )
+            to_save["result"] = data
+
+        statement = postgres_upsert(TaskResult).values([copy.copy(to_save)])
+        del to_save["id"]
+        statement = statement.on_conflict_do_update(index_elements=[TaskResult.id], set_=to_save)
```

Comment:
I am not sure about this bit: I know you want to provide upsert capability to postgres as per docs, but I don't get why you are deleting `id` key

Reply:
I don't want to set the ID, as it's already set.